Tests/SwiftGRPCNIOTests/NIOClientTimeoutTests.swift

```diff
@@ -0,0 +1,132 @@
+import SwiftGRPCNIO
+import NIO
+import XCTest
+
+class NIOClientTimeoutTests: NIOBasicEchoTestCase {
+  static var allTests: [(String, (NIOClientTimeoutTests) -> () throws -> Void)] {
+    return [
+      ("testUnaryTimeoutAfterSending", testUnaryTimeoutAfterSending),
+      ("testServerStreamingTimeoutAfterSending", testServerStreamingTimeoutAfterSending),
+      ("testClientStreamingTimeoutBeforeSending", testClientStreamingTimeoutBeforeSending),
+      ("testClientStreamingTimeoutAfterSending", testClientStreamingTimeoutAfterSending),
+      ("testBidirectionalStreamingTimeoutBeforeSending", testBidirectionalStreamingTimeoutBeforeSending),
+      ("testBidirectionalStreamingTimeoutAfterSending", testBidirectionalStreamingTimeoutAfterSending)
+    ]
+  }
+
+  private func expectDeadlineExceeded(forStatus status: EventLoopFuture<GRPCStatus>) {
+    let statusExpectation = self.expectation(description: "status received")
+
+    status.whenSuccess { status in
+      XCTAssertEqual(status.code, .deadlineExceeded)
+      statusExpectation.fulfill()
+    }
+
+    status.whenFailure { error in
+      XCTFail("unexpectedelty received error for status: \(error)")
+    }
+  }
+
+  private func expectDeadlineExceeded(forResponse response: EventLoopFuture<Echo_EchoResponse>) {
+    let responseExpectation = self.expectation(description: "response received")
+
+    response.whenFailure { error in
+      XCTAssertEqual((error as? GRPCStatus)?.code, .deadlineExceeded)
+      responseExpectation.fulfill()
+    }
+
+    response.whenSuccess { response in
+      XCTFail("response recevied after deadline")
+    }
+  }
+}
+
+extension NIOClientTimeoutTests {
+  func testUnaryTimeoutAfterSending() {
+    // The request gets fired on call creation, so we need a very short timeout.
+    let callOptions = CallOptions(timeout: .milliseconds(1))
+    let call = client.get(Echo_EchoRequest(text: "foo"), callOptions: callOptions)
+
+    self.expectDeadlineExceeded(forStatus: call.status)
+    self.expectDeadlineExceeded(forResponse: call.response)
+
+    waitForExpectations(timeout: defaultTimeout)
+  }
+
+  func testServerStreamingTimeoutAfterSending() {
+    // The request gets fired on call creation, so we need a very short timeout.
+    let callOptions = CallOptions(timeout: .milliseconds(1))
+    let call = client.expand(Echo_EchoRequest(text: "foo bar baz"), callOptions: callOptions) { _ in }
+
+    self.expectDeadlineExceeded(forStatus: call.status)
+
+    waitForExpectations(timeout: defaultTimeout)
+  }
+
+  func testClientStreamingTimeoutBeforeSending() {
+    let callOptions = CallOptions(timeout: .milliseconds(50))
+    let call = client.collect(callOptions: callOptions)
+
+    self.expectDeadlineExceeded(forStatus: call.status)
+    self.expectDeadlineExceeded(forResponse: call.response)
+
+    waitForExpectations(timeout: defaultTimeout)
+  }
+
+  func testClientStreamingTimeoutAfterSending() {
+    let callOptions = CallOptions(timeout: .milliseconds(50))
```

Comment:
How about 10/20ms timeout and sleeping 11/21, for faster test runtimes? (Also elsewhere)